[DOC] Fix errors in the description for installation on Windows

emankov opened the pull request:
+ All quotes for ENV VARS are erroneous;
+ Toolset hasn't be specified;
+ Provide paths for all 3 Visual Studio 2017 products: Community/Professional/Enterprise.

emankov committed:
[DOC] Fix errors in the description for installation on Windows

+ All quotes for ENV VARS are erroneous;
+ Toolset hasn't be specified;
+ Provide paths for all 3 Visual Studio 2017 products: Community/Professional/Enterprise.

soumith commented:
cc: @peterjc123 does this look good?

peterjc123 commented on README.md:
````diff
@@ -184,16 +184,21 @@ MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
 On Windows
 ```cmd
-set "VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build"
+REM The folowing line is for Visual Studio 2017 Enterprise only
````
You missed `BuildTools`. It is also a product of VS 2017. But I don't think listing products here is a good idea. Actually, you can use `vswhere`. The usage is listed here: https://github.com/peterjc123/pytorch-scripts/blob/master/internal/check_deps.bat#L17-35. It is also acceptable if you just list them in the comments.

emankov replied:
> You missed `BuildTools`

By `BuildTools` do you mean
`VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\Tools\`
which is set after vcvarsall.bat invocation?

As for `vswhere` for Visual Studio 2017 only, well it is not really useful for the end user as it outputs a big list of info and that's it. But you will still need to mention in Readme `..\VC\Auxiliary\Build` as it is not presented in its output. Nonetheless, it might be used for obtaining `installationPath`, to which `\VC\Auxiliary\Build` should be appended (manual copy-paste, IMHO).

peterjc123 replied:
> By `BuildTools` do you mean
`VS150COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\Tools\`
which is set after vcvarsall.bat invocation?

No, I mean `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build`. It can be installed using the standalone installer. Please read more details [here](https://devblogs.microsoft.com/cppblog/introducing-the-visual-studio-build-tools/).

> As for `vswhere` for Visual Studio 2017 only, well it is not really useful for the end user as it outputs a big list of info and that's it. But you will still need to mention in Readme `..\VC\Auxiliary\Build` as it is not presented in its output. Nonetheless, it might be used for obtaining `installationPath`, to which `\VC\Auxiliary\Build` should be appended (manual copy-paste, IMHO).

Please read the script carefully (in the link). It also does what you said. VS 2017 isn't a problem because our code relies on it.

emankov replied:
1. I saw it in your script, but it looks like it is not presented in pytorch? Correct me if I wrong.
2. The name of the 'product' (Enterprise) has been already in Readme. I've just added the rest 2 (Community and Professional), and that's it. Otherwise, it looked like Enterprise is the only suitable 'product'.
So, what you want me to correct here? I didn't test with [Build tools](https://devblogs.microsoft.com/cppblog/introducing-the-visual-studio-build-tools/), so I couldn't write anything about it here.

peterjc123 replied:
1. Yes, it is. It's not in because it is somehow too complex.
2. Instead of listing the paths here, just list the products here. And please add in `BuildTools`, whose path structure is just the same as the other products.

emankov replied:
1. Maybe a single command will suit us?

``FOR /f "usebackq tokens=*" %i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -property installationPath`) do SET VS150COMNTOOLS=%i\VC\Auxiliary\Build``

Additionally, we may ask a user to check: `echo %VS150COMNTOOLS%` or may not, cause the further call will check it: `call "%VS150COMNTOOLS%\vcvarsall.bat" x64`

emankov replied:
2. Will experiment with BuildTools and VC++ 2017 v141 toolset, but would prefer to mention it in Readme in another PR.

peterjc123 replied:
Sounds good.

emankov replied:
1. Let's cram it in a single command:

``FOR /f "usebackq tokens=*" %i IN (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -property installationPath`) DO CALL "%i\VC\Auxiliary\Build\vcvarsall.bat" x64``

I've tested it from Command Prompt on Windows 7, 8, 8.1 and 10. Where both Visual Studio 2015 and 2017 were installed (Community and Professional).

peterjc123 replied:
looks okay to me

peterjc123 replied:
Wait, you 'd better add the argument `-products *` into this, which will also find buildtools.

emankov replied:
Are you sure about build tools? It looks like it is about products only.
```
  -products arg  One or more product IDs to find. Defaults to Community, Professional, and Enterprise.
                 Specify "*" by itself to search all product instances installed.
                 See https://aka.ms/vs/workloads for a list of product IDs.
```
Btw, the above `for` command should be testes on multiple Visual Studio 2017 installed. I'm going to install Community to check such config.

peterjc123 replied:
I've tested this argument in real CI. Without it, it will only find `Community, Professional, and Enterprise` products.

emankov replied:
Could you please provide your output, because I've just successfully installed "VC++ 2017 version 15.4 v14.11 toolset", and adding `-products *` doesn't change anything in my case:

```
C:\Windows\system32>"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -products * -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional
```

peterjc123 replied:
```cmd

C:\Users\Administrator\Desktop>"%ProgramFiles(x86)%\Microsoft Visual Studio\Inst
aller\vswhere.exe" -version [15^,16^) -products * -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community

C:\Users\Administrator\Desktop>"%ProgramFiles(x86)%\Microsoft Visual Studio\Inst
aller\vswhere.exe" -version [15^,16^) -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community

C:\Users\Administrator\Desktop>
```

emankov replied:
Hmmm... I See only "Professional", and there is no corresponding folder for BuildTools near it as well, even though I successfully installed "VC++ 2017 version 15.4 v14.11 toolset". Have I missed something concerning BuildTools?

emankov replied:
![02](https://user-images.githubusercontent.com/10241941/53571648-79fa3380-3b7a-11e9-8098-1e7b2ae9a595.png)

peterjc123 replied:
No no no, by buildtools, I mean installing VS through this installer `https://download.visualstudio.microsoft.com/download/pr/f8db94a7-42dc-4413-b987-b21dcb0b1661/5e24cf2c187f46bde25c5515250fa9a0/vs_buildtools.exe`. It is also a product, just like `Professional` and it can be installed even if you have some other products like `Professional`.

emankov replied:
It looks like it is not shipped by Microsoft currently... Well, nonetheless, do we need BuildTools to be installed and mentioned in Readme? Or that is yet another Visual Studio Product like Community/Professional/Enterprise which might be used for building pytorch?

peterjc123 replied:
> It looks like it is not shipped by Microsoft currently...

It's actually shipped by Microsoft in https://visualstudio.microsoft.com/downloads/.
![image](https://user-images.githubusercontent.com/9998726/53573038-99f41c00-3ba7-11e9-8a66-8e165ad546df.png)

> Or that is yet another Visual Studio Product like Community/Professional/Enterprise which might be used for building pytorch?

When using `-products *`, we are using all the possible products. So we don't have to care about it.

emankov replied:
Ok, thanks, I missed BuildTools on Microsoft.com.
Now we should choose, I think, one product in order not to get multiple environment initialisations in the loop.

```
C:\Windows\system32>for /f "usebackq tokens=*" %i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [14^,16^) -products * -property installationPath`) do call "%i\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11

C:\Windows\system32>call "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Windows\System32>call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.11
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
```

peterjc123 replied:
Agreed.

emankov replied:
I think only the user can choose the product in case of multiple products installed. Searching an appropriate product with the needed toolset is the work for the script, not for a couple of commands. We can check such situation with multiple products and let him/her choose, otherwise (if only one VS 2017 product is installed) do the job seamlessly without user engagement. What do you think?

emankov replied:
One more option is to use "-latest":

`-latest        Return only the newest version and last installed.`

In this case, a single product will always be chosen. If the product will not suit - the user will have to update it and/or install the needed toolset.

I like this option more.

peterjc123 commented on README.md:
```diff
@@ -184,16 +184,21 @@ MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
 set FORCE_PY27_BUILD=1
 REM As for CUDA 8, VS2015 Update 3 is also required to build PyTorch. Use the following line.
-set "CUDAHOSTCXX=%VS140COMNTOOLS%\..\..\VC\bin\amd64\cl.exe"
+set CUDAHOSTCXX=%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe
```
Why do you remove the quotes? According to this [article](https://ss64.com/nt/set.html), although it looks a bit weird, it is the right thing to do, because we should put the quotes here when there is space in the string when using set.

emankov replied:
Because they are erroneous. Well, ok, to be honest, they are definitely erroneous for all VS env. vars like VS150COMNTOOLS. Here you may put quotes, but only this way:
`CUDAHOSTCXX="%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe"`

peterjc123 replied:
What error does it throw? Are you using Git Bash, MinGW shell or CMD?

emankov replied:
Far

peterjc123 replied:
Well, I tested and the problem exists. But I don't think we should remove the quotes here. They are working in CMD. It is the problem of this console emulator. I tested the command on some other emulators like `ConEmu` and it also worked. It's not acceptable that we remove the quotes here only for one console emulator, which will then cause problems for many users that use CMD.

emankov replied:
`set CUDAHOSTCXX="%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe"`
And there will no problems at all.

peterjc123 replied:
Result in CMD:
```cmd
Microsoft Windows [Version 10.0.17763.316]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Users\peter>set CUDAHOSTCXX="%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe"

C:\Users\peter>echo %CUDAHOSTCXX%
"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\..\..\VC\bin\amd64\cl.exe"

C:\Users\peter>set "CUDAHOSTCXX_OLD=%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe"

C:\Users\peter>echo %CUDAHOSTCXX_OLD%
C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\..\..\VC\bin\amd64\cl.exe
```
Apparently, the quotes are added to the variable if I choose your option, which won't happen in the old one.

emankov replied:
It is ok, it works, it is a convenient common practice to enclose in quotation marks variable values, in which spaces might be presented. Compilers do it, for instance.

peterjc123 replied:
It works, but not in a safe way. Let me explain it in detail. When `set "c=XXXXXXXXXXXXXXX"` is used, the quotes here are called protection quotes. It will prevent anything in the string to be interpreted as something else. We've received quite a lot of issues about this, which will complain that `\Microsoft is unexpected here`. If I choose your option, then these problems will occur again.

emankov replied:
Ok, let's leave it as is, moreover, `cmd` should be called beforehand.

peterjc123 commented on README.md:
```diff
@@ -184,16 +184,21 @@ MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install
+set CUDAHOSTCXX=%VS140COMNTOOLS%..\..\VC\bin\amd64\cl.exe
 
-call "%VS150COMNTOOLS%\vcvarsall.bat" x64 -vcvars_ver=14.11
+call "%VS150COMNTOOLS%\vcvarsall.bat" x64
```
Well, if you remove the 14.11 constraint here, then you should add a comment here. Different versions of CUDA rely on different versions of VS. And I can only confirm that (VS 15.3/ VC++14.11)  work for major CUDA versions that we use. (CUDA 9 - CUDA 10)

emankov replied:
> if you remove the 14.11 constraint here

It is not actually a constraint. With that option user will get the following error:
`[ERROR:vcvars.bat] Toolset directory for version '14.11' was not found.`
for all versions of VS 2017, except Update 3 and 4 (15.3.3, 15.4.4 and 15.4.5). Here is a more or less complete list of all Visual Studio versions (unofficial): [Visual Studio versions](https://dev.to/yumetodo/list-of-mscver-and-mscfullver-8nd). I believe that he/she in case of the above error will try the same as I did: `-vcvars_ver=14.1x`, where `x` is his/her minor toolset version. Finally, I removed this option and built successfully by VS 2017 (15.9.5, toolset 14.16).

> And I can only confirm that (VS 15.3/ VC++14.11) work for major CUDA versions that we use. (CUDA 9 - CUDA 10)

It should be tested with the rest of VS 2017 (15.x) versions and CUDA 9.x-10.0. If the constraint actually takes place - it is bad news, because Visual Studio is being updated actively and there is no mechanism for uninstalling its updates. After that such constraints should be mentioned above in Readme somewhere near "If you want to build on Windows..." or section 'Constraints' should be added.

peterjc123 replied:
Sounds reasonable. Would you please add it? Some users have tested the conditions for CUDA 9 and CUDA 9.2. https://github.com/pytorch/pytorch/issues/15630#issuecomment-451247858

emankov replied:
Sorry, what you mean by 'it'? What you think should be added in Readme?

emankov replied:
I didn't test those configurations so I couldn't guarantee them. Moreover, I have a multiple CUDA  (7.0/7.5/8.0/9.0/9.1/9.2/10.0), and multiple VS (14.0 Update 3 Pro/15.9.5 Pro) installed, and still do not know which CUDA was chosen by pytorch... I'm a new "forced" user of your framework. And what I may guarantee is that the diff in this PR works for me entirely.

The only thing that I may say on the subject is that better to stay in Readme in terms of versions (like 15.x.x), which the user may check easily, than in terms of macros' values. But in your build scripts, you may (but I'd prefer 'should') add your macro based constraints with corresponding warnings and/or errors.

peterjc123 replied:
> I'm a new "forced" user of your framework. And what I may guarantee is that the diff in this PR works for me entirely.

Yes, but as a developer here, I have to consider for most users. 

> I didn't test those configurations so I couldn't guarantee them.

I pick these combinations out from the code `%CUDA_PATH%\include\crt\host_defines.h`, so they are absolutely true.

> Moreover, I have a multiple CUDA (7.0/7.5/8.0/9.0/9.1/9.2/10.0), and multiple VS (14.0 Update 3 Pro/15.9.5 Pro) installed, and still do not know which CUDA was chosen by pytorch (looks from the build log like the latest 10)

It will depend on your env vars like `PATH` or `CUDA_PATH`, Usually, it will pick the last one your install.

emankov replied:
CUDA_PATH should be mentioned in Readme as well.

peterjc123 reviewed (changes requested):
Thanks for the PR. The review notes are listed below.

emankov committed:
Merge branch 'master' of https://github.com/pytorch/pytorch

emankov committed:
[DOC] Update the description for installation on Windows

+ Describe Visual Studio 2017 constraints
+ Describe NVIDIA Tools Extension (NVTX)
+ Update commands to run
+ Move all Windows related text under "On Windows" section

peterjc123 commented:
LGTM, just some.nits: would it be better if you replace these `REM`s with `::`?
